### ISSUE TITLE
(core) - Unknown cause of memory leak in Node (TBD)

### ISSUE THREAD

**Issue by rvalle** [urql-graphql/urql#2355]
urql version & exchanges: urql/core 2.4.3 + cross-fetch 3.1.5 in NodeJS.

**Steps to reproduce**

1. Create a simple query with toPromise()
2. Place it in an infinite loop
3. Watch memory growth 

**Expected behavior**

garbage collector cleans up old objects

**Actual behavior**

Closures, Request, Wonka objects remain in memory

After realizing issue #2353 was not related to Caching, I started removing code from my data processing application, until, literally just having a simple query in a loop. The Memory leak would still take place.

Here I attach memory charts as seen production, followed by memory after migrating the code to an alternative Graphql client with almost compatible interface:

![Indexer Monitoring - Grafana](https://user-images.githubusercontent.com/412837/158963160-7f5f94e9-9171-40bb-8937-86534b74c915.png)

**Comment by JoviDeCroock** [comment]
Hey as referenced in that previous issue have you tried with v2.0.0 and 1.16.2, maybe we can track down whether this has always been there or is caused by an addition in urql.

**Comment by rvalle** [comment]
I think I did, not entirely sure about 1.16.2, but definitely tried all the 2.x latest minors and the behavior seemed identical.

**Comment by JoviDeCroock** [comment]
I mean, we might need a bit more to work with, like a way to reproduce this or notice this 😅 there might be a lot going on in your code that makes this irreproduceable for us on a clean core

**Comment by rvalle** [comment]
sorry @JoviDeCroock I have already migrated to another toolkit.

In my last tests there was nothing but a simple query with .toPromise() in a loop, getting 10 rows on each execution. And not doing anything with the returned result.

Normally I put more effort on contributing but closing the previous issue despite letting know that I would come later with the result of the cache refactoring prompted me to look for an alternative lib, also... comments are disabled on closed issues too, I found out after writing a follow up.

**Comment by frederikhors** [comment]
> Normally I put more effort on contributing but closing the previous issue despite letting know that I would come later with the result of the cache refactoring prompted me to look for an alternative lib, also... comments are disabled on closed issues too, I found out after writing a follow up.

I tend to stick to this idea. There is no point in closing an issue IMMEDIATELY if the person who found a problem is working hard to understand the cause.

@rvalle, may I ask which alternative did you find?

**Comment by JoviDeCroock** [comment]
@rvalle You were still able to comment afaik, I mean just knowing how your client looked could go a long way for us but yes wishing you the best of luck anyway

**Comment by rvalle** [comment]
@frederikhors for my application ```graphql-request``` is enough. I don't use UI features, mutations, or other things. I could have used fetch/post also....

@JoviDeCroock I do appreciate your project and contribution to opensource. I was not even upset, it is just that I also have a delivery calendar and I am in a rush. 

don't know what happened with the other issue but after writing a long follow up is said "you cannot comment at this time".... I made a quick test now, and I its actually allowing comments. So, might have being some kind of glitch. 

But honestly, I think the issue should be very easy to reproduce. Let me know if you cannot.

**Comment by kitten** [comment]
@rvalle We've got limited time and hence it's commonplace for reproductions to be necessary for non-obvious issues or questions to be resolved, meaning, if it's not simply in an existing example we provide then it takes work for us to make it. So, it's unfortunately not a question of whether it's easy or not to reproduce, but if you've got a code sample that reproduces the issue you achieve a couple of things that move all of us forward:

- You're convincing yourself of the minimum (or close to minimum) requirements that are needed to reproduce the issue
- You're providing a realistic scenario in case we need to make API changes or to test against (which makes it also easier to close an issue, even if you ran out of time to come back to us here)
- We can see whether there's a potential misunderstanding between us before we go out and try to reproduce the exact sample of code you may have written (which eliminates assumptions)

So a code reproduction saves exponentially more time with each of these points.
That said, if you can also just post code here that you ran, I can see first whether there's something obvious we've missed or that causes this.

On the issue of commenting: closed issues are still open for commenting unless they're locked. GitHub did have API issues on the day you tried to potentially reply on though. So, that's likely what's caused this.

On the note that we closed the last issue on: The `cacheExchange` is not an LRU cache and if you're using `@urql/core` for one off requests you'll have to override the `exchanges` options and _not_ pass a cache exchange, i.e. `exchanges: [dedupExchange, fetchExchange]`

Now, I assume you've done this, but hence what is actually retaining references to the responses becomes hard to tell for me. I had a scan through our `Client` a couple of days ago to rule out obvious mistakes from our end that'd maybe cause our "shared response sources" from being GC'd, but couldn't see anything obvious there

**Comment by kitten** [comment]
We think this may have been caused by a specific usage pattern with `toPromise` where multiple identical queries could leave behind copies of their result: https://github.com/FormidableLabs/urql/pull/2386

I'll close this for now, since I think this was the only condition that couldve caused this ✌️

**Comment by ShakedHad** [comment]
> We think this may have been caused by a specific usage pattern with `toPromise` where multiple identical queries could leave behind copies of their result: #2386
> 
> I'll close this for now, since I think this was the only condition that could've caused this ✌️

Hi
Unfortunately, I'm experiencing the exact situation that you are describing here, 
I'm using @urql/core 3.0.3 to perform a lot of one-off requests (using toPromise), at a node express server, when some of the requests could be heavy (around 10MB).

Examining memory dump diffs shows undisposed Wonka objects:
![image](https://user-images.githubusercontent.com/25921813/196688036-cd64e461-7745-4de2-b55e-519f59d4b7d5.png)

Have you encountered any more reports like those? is it possible this bug wasn't fixed successfully

**Comment by wernst** [comment]
Anecdotally I am also seeing memory creep up as requests are sent with `toPromise` in a React Native app using @urql/core 3.0.3. I will try to examine more thoroughly, but as a simple binary test removing the requests does seem to "fix" the leak.

**Comment by shevchenkonik** [comment]
Hey folks, I can't confirm that the problem has been solved, I don't understand why the issue was closed. The same issue is still represented for me with the latest version of `@urql/core`. In meantime, I use the correct settings for exchanges which include only `fetchExchange` to prevent some caching there, and `requestPolicy:  'network-only'`.

**Comment by JoviDeCroock** [comment]
Can anyone reproduce this? Like just a sandbox and some steps 😅

**Comment by shevchenkonik** [comment]
@JoviDeCroock 

I couldn't share our production codebase or metrics but I've reproduced it as a part of a simple express application and I found the problem reproduced in the situation when I use the `@urql/core` client as a singleton. In case when I create a client for each request everything looks more stable and without memory leaking. 

As @rvalle proposed, we can reproduce it with a simple loop or local DDOS (30+ RPS is fine) of the express app:

```typescript
import 'isomorphic-unfetch';
import * as express from 'express';
import { createClient, fetchExchange, gql } from '@urql/core';

const fs = require('fs');
const v8 = require('v8');

function createHeapSnapshot() {
    const snapshotStream = v8.getHeapSnapshot();
    const fileName = `${Date.now()}.heapsnapshot`;
    const fileStream = fs.createWriteStream(fileName);
    snapshotStream.pipe(fileStream);
}

const app = express();
const port = 3000;

const query = gql`
    query {
        countries {
            name
        }
    }
`
const urqlClient = createClient({
     url: 'https://countries.trevorblades.com/',
     exchanges: [fetchExchange],
     requestPolicy: 'network-only',
});

app.use(require('express-status-monitor')());

app.get('/', async(req, res) => {
    await urqlClient.query(query, { test: new Date() }).toPromise().then()
    res.send('Hello World!')
});

const makeHeapSnapshot = () => {
    setInterval(() => {
        createHeapSnapshot()
    }, 20000)
}

makeHeapSnapshot()

app.listen(port, () => console.log(`Server started listening to port ${port}`));
```

![Screenshot 2022-11-23 at 16 39 16](https://user-images.githubusercontent.com/8392253/203587717-d114f720-edac-4afd-b0e2-f4ea5a9d171a.png)

You can see that memory is continiously growing:
![image](https://user-images.githubusercontent.com/8392253/203588751-1d81c047-b9b3-4464-9dc5-b194a055d789.png)


If you move the initialization of the client to request, it will work fine (client isn't a singleton)

```typescript
app.get('/', async(req, res) => {
    const urqlClient = createClient({
         url: 'https://countries.trevorblades.com/',
         exchanges: [fetchExchange],
         requestPolicy: 'network-only',
    });
    await urqlClient.query(query, { test: new Date() }).toPromise().then()
    res.send('Hello World!')
});
```

![Screenshot 2022-11-23 at 16 40 06](https://user-images.githubusercontent.com/8392253/203587880-22d453dc-c3b3-4b58-9514-6bc40d3c5943.png)

You can see that memory is fine:
![Screenshot 2022-11-23 at 16 41 05](https://user-images.githubusercontent.com/8392253/203588088-f9fb20b5-0b4d-48da-b51c-aca6979e8696.png)

Also I have the similar heap snapshot as @ShakedHad attached.

I din't go deeper to codebase for finding the reason yet but I hope this quick research will be helpful for you.

**Comment by kitten** [comment]
Just stepping in here really quick to say that we'll need details beyond these. We're very clear on what people are doing, however, in this case a heap snapshot (e.g. Chrome's devtools should be able to export memory profiles as per the snapshots shown) would be very helpful here.

In short, we'd need some details on the traced, retained objects to actually be able to see why this is happening, because we haven't quite been able to reproduce this cleanly, i.e. we haven't seen consistently retained objects and instead what I've seen is that usually retained "replay objects" (which is a secondary mini-cache in the `Client`) disappear cleanly 🤔

**Comment by ShakedHad** [comment]
Hi @kitten 
Thanks for your response, regarding your note on when to init the client:
I have to say it is really unclear from the documentation that these are your intentions (at least for me), specifically this part (https://formidable.com/open-source/urql/docs/basics/core/#one-off-queries-and-mutations)

**Comment by kitten** [comment]
@ShakedHad: To be fair, there's a difference here between SSR and non-SSR. What I mean is, if we were to make thousands of requests on one `Client`, we of course don't expect to create any memory leak. However, I don't want anyone to walk away with the wrong expectations here in case it's an unrelated issue 😅

Currently, I mostly expect that there's something wrong with the promise creation, but I don't have enough of an idea to go on here yet, since we previously already fixed a leak associated with that.

However, we don't have any relevant version info (`@urql/core` mainly), memory profiles that point at some objects in the stack, and such, which makes it hard to trace this down.
In other words, while this issue has collected a few new replies, every single case may in fact be different and only have Node.js in common. That's of course fine, but it's not the Node.js environment itself that's causing the leak, I mean, so our "original" issue creation requirements apply and we'll need new info to look into this :v:

**Comment by jangot** [comment]
So I have reproduced the problem and have created an [example](https://github.com/jangot/urql-memory) for you.
You need to start 2 servers and start the script for calling those.
<img width="1439" alt="Screenshot 2022-11-24 at 12 20 04" src="https://user-images.githubusercontent.com/1309552/203772364-6d2382e2-163d-4ccf-a6f8-998e9a16f595.png">

You can see on the screenshot that after 1000 requests it starts consume more than 160mb and it will not be clean after finishing the script.

**Comment by jangot** [comment]
And also I created an example where I created `urql` for every incoming request.
so I suggest you to start first test:

1. Open terminal and run this: `npm run gql-server`
2. Open second terminal and run this: `npm run request-server`
3. Open third terminal and tun: `npm run start`
4. Open `http://localhost:3000/status` and check memory

Next time I suggest you to change second step and run `npm run request-uniq-server`.
You will see on `http://localhost:3000/status` that memory doesn't go up.

**Comment by kitten** [comment]
Just giving an update here; This will be resolved with the next release ✌️

It's been really hard to chase this issue down, since it was in a very unrelated branch of the `fetchExchange` hidden in a stream operator. We believe the regression snuck in when we migrated that codebase over to TypeScript, since it came down to basically a single missing line of code / a single call.

While we spent a lot of time hunting this down, and it did take longer since the leak was small and steady, we did also take the opportunity to take a look into an alternative implementation of the `fetch` logic we had, which I worked on a while ago.

It's a more maintainable version of our current code and is based on async generators. Those should actually help the newer versions of Node.js/V8 a little along when it's under more load. At least, I've observed some small improvements while I was hunting down this issue.

However, that change will be delayed, so we have enough time to first get the memory leak patch shipped, without having to worry about introducing new regressions accidentally ✌️
